tests/unit/test_command_install.py

```diff
@@ -9,39 +10,90 @@
-            (True, False, True),
-            (False, True, False),
-            (False, False, False),
-        ])
-    def test_most_cases(
-        self, enable_user_site, site_packages_writable, result, monkeypatch,
-    ):
-        monkeypatch.setattr('site.ENABLE_USER_SITE', enable_user_site)
-        monkeypatch.setattr(
-            'pip._internal.commands.install.site_packages_writable',
-            lambda **kw: site_packages_writable
-        )
-        assert decide_user_install(use_user_site=None) is result
+    @pytest.mark.parametrize('use_user_site,prefix_path,target_dir,root_path',
+                             filter(lambda args: sum(map(bool, args)) > 1,
+                                    product((False, True), (None, 'foo'),
+                                            (None, 'bar'), (None, 'baz'))))
+    def test_conflicts(self, use_user_site,
+                       prefix_path, target_dir, root_path):
+        with pytest.raises(CommandError):
+            decide_user_install(
+                use_user_site=use_user_site, prefix_path=prefix_path,
+                target_dir=target_dir, root_path=root_path)
+
+    def test_target_dir(self):
+        with patch(WRITABLE, true):
+            with patch(EXISTS, true), patch(ISDIR, false):
+                with pytest.raises(InstallationError):
+                    decide_user_install(target_dir='bar')
+            for exists, isdir in (false, false), (false, true), (true, true):
+                with patch(EXISTS, exists), patch(ISDIR, isdir):
+                    assert decide_user_install(target_dir='bar') is False
+
+    def test_target_writable(self):
+        with patch(EXISTS, false):
+            with patch(WRITABLE, false), pytest.raises(InstallationError):
+                decide_user_install(target_dir='bar')
+            with patch(WRITABLE, true):
+                assert decide_user_install(target_dir='bar') is False
+
+    def test_prefix_writable(self):
+        with patch(WRITABLE, false), pytest.raises(InstallationError):
+            decide_user_install(prefix_path='foo')
+        with patch(WRITABLE, true):
+            assert decide_user_install(prefix_path='foo') is False
+
+    def test_global_site_writable(self):
+        with patch(ENABLE_USER_SITE, True):
+            with patch(WRITABLE, false):
+                with pytest.raises(InstallationError):
+                    decide_user_install(use_user_site=False)
+                with pytest.raises(InstallationError):
+                    decide_user_install(root_path='baz')
+                assert decide_user_install() is True
+            with patch(WRITABLE, true):
+                assert decide_user_install(use_user_site=True) is True
+                assert decide_user_install(root_path='baz') is False
+                assert decide_user_install() is False
+
+    def test_enable_user_site(self):
+        with patch(WRITABLE, true):
+            with patch(ENABLE_USER_SITE, None):
```

Comment:
It's recommended to use `monkeypatch` fixture with pytest, not a regular patch decorator.

Reply:
Thanks, I saw `patch` being imported and thought it's conventionally encouraged.